### ISSUE TITLE
`unnecessary_cast` false positive

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/linter#4277]
**Describe the issue**
I have a function with a generic type `T extends Object?`. Somewhere in the middle of my function, I have a test to see if my variable `T object` is `DateTime?`. I'm not sure how to describe this better than showing you the code (it can be tested with dartpad.dev:

**To Reproduce**
```dart
import 'package:intl/intl.dart';

/// This is an extension method so I don't always have to test for nullable values*
extension Apply<T extends Object> on T {
  R apply<R>(R Function(T self) transform) => transform(this);
}

void func<T extends Object?>(T object) {
  if (object is DateTime?) {
    final formatter = DateFormat('yyyy/MM/dd');
    object?.day; // No lint
    if (object != null) {
      formatter.format(object); // No lint
    }
    object?.apply(formatter.format); // The argument type 'String Function(DateTime)' can't be assigned to the parameter type 'dynamic Function(T)'.
    (object as DateTime?)?.apply(formatter.format); // Unnecessary cast.
  }
}
```
*[Here is where I found it](https://github.com/dart-lang/language/issues/361#issuecomment-1347411411)

**Expected behavior**
Either not triggering `unnecessary_cast` or not triggering the error with the line above as all the other code before works just fine.

**Comment by srawlins** [comment]
I am more surprised about the error at `object?.apply(formatter.format);`. CC @scheglov 

I'm not sure why inference isn't setting `T` to `DateTime` there. `object` is confirmed to be promoted to `DateTime?` as per the lines like `object?.day`, and the receiver of `object?.apply` should be promoted to `DateTime`, I think...

**Comment by scheglov** [comment]
Yeah, type inference is weird in this invocation.
See below what we have currently, but don't consider it to be the right expectation.
Apparently we don't infer `R` correctly.
```dart
  solo_test_X() async {
    await assertErrorsInCode(r'''
void f<T extends Object?>(T object) {
  if (object is int?) {
    object?.apply(g);
  }
}

extension E<T2 extends Object> on T2 {
  void apply<R>(R Function(T2 self) transform) {}
}

double g(int a) => 0.0;
''', [
      error(CompileTimeErrorCode.ARGUMENT_TYPE_NOT_ASSIGNABLE, 80, 1),
    ]);

    final node = findNode.singleMethodInvocation;
    assertResolvedNodeText(node, r'''
MethodInvocation
  target: SimpleIdentifier
    token: object
    staticElement: self::@function::f::@parameter::object
    staticType: T & int?
  operator: ?.
  methodName: SimpleIdentifier
    token: apply
    staticElement: MethodMember
      base: self::@extension::E::@method::apply
      substitution: {T2: T, R: R}
    staticType: void Function<R>(R Function(T))
  argumentList: ArgumentList
    leftParenthesis: (
    arguments
      SimpleIdentifier
        token: g
        parameter: ParameterMember
          base: root::@parameter::transform
          substitution: {R: dynamic}
        staticElement: self::@function::g
        staticType: double Function(int)
    rightParenthesis: )
  staticInvokeType: void Function(dynamic Function(T))
  staticType: void
  typeArgumentTypes
    dynamic
''');
  }
```

**Comment by srawlins** [comment]
Looks like the same bug as: https://github.com/dart-lang/sdk/issues/35799

**Comment by FMorschel** [comment]
Please give a look at this comment on [dart-lang/sdk#52077](https://github.com/dart-lang/sdk/issues/52077#issuecomment-1513225793)

**Comment by pq** [comment]
Moving this one to the SDK tracker since `UNNECESSARY_CAST` is an analyzer warning code.
